### PR TITLE
[PXSOD-9480] Start service in foreground per API reqs

### PR DESCRIPTION
**Pull request by ssolbeck**
From https://developer.android.com/guide/components/services#StartingAService

> If your app targets API level 26 or higher, the system imposes restrictions on using or creating background services unless the app itself is in the foreground. If an app needs to create a foreground service, the app should call [startForegroundService()](https://developer.android.com/reference/android/content/Context#startForegroundService(android.content.Intent)). That method creates a background service, but the method signals to the system that the service will promote itself to the foreground. Once the service has been created, the service must call its [startForeground()](https://developer.android.com/reference/android/app/Service#startForeground(int,%20android.app.Notification)) method within five seconds.

### CODE REVIEW

**Comment by ssolbeck** [comment]
I'm concerned that with this change that it may cause breakages at locations which have _not_ updated the PrintConnect app -- it seems that if the service being started does not declare itself foregrounded, the service-intent will fail. It's not clear to me whether PrintConnect called `startForeground` previously.

That being said, since GPM is the only merchant using Zebra, perhaps the requirement that both apps be updated is acceptable.

**Review by bsoper — COMMENTED** [review]
This seems reasonable to me since GPM is the only one using it. Hopefully they have upgraded all locations though. I know we considered suggesting that they try to downgrade the PrintConnect app version, so if they have done so, we may end up with different stores on different versions.
Do we know if there is a way to detect which version of the print connect app they are using so we can handle both flows?

**Comment by ssolbeck** [comment]
> Do we know if there is a way to detect which version of the print connect app they are using so we can handle both flows?

Yes, seems like it would be possible -- but involves some unknowns (what permissions are necessary, what format the version will come through in) that may be hard to test with our limited Android setup.

https://developer.android.com/training/package-visibility

**Review by bsoper — APPROVED** [review]
Understood. I am approving with the assumption that we can tell GPM that a certain version of our PrintPlugin and a certain version of PrintConnect are required for this to work.

**Review by jacobjanak — APPROVED** [review]
Whoops, forgot to hit approve

**Comment by ssolbeck** [comment]
Going to wait on UAT from Travis at GPM -- once we have the go-ahead there, I'll simply merge this to `master`, then bump the version on the downstream order management app, which will use the latest version of this plugin.

File: src/android/Print.java
Comment (bsoper) [inline src/android/Print.java]: Is it guaranteed for this package info to exist? If not, we should check that it is populated before calling `getLongVersionCode()`

File: src/android/Print.java
Comment (bsoper) [inline src/android/Print.java]: What is this a TODO for?

Comment (ssolbeck) [reply src/android/Print.java]: Testing purposes -- this 0 should be the actual version threshold, but I don't know what that should be yet.

File: src/android/Print.java
Comment (ssolbeck) [reply src/android/Print.java]: It will throw an exception if not found -- included handling for that now.

File: src/android/Print.java
Comment (jacobjanak) [inline src/android/Print.java:728]: I think this code will break when version 1.3.0 is released

Rather than handling that case via code, it might be easier to have a follow up PR for once GPM is all on the newest version of printconnect. After they're all updated, we can just remove all the version checking stuff and always do the foreground approach

**Review by jacobjanak — APPROVED** [review]

File: src/android/Print.java
Comment (ssolbeck) [reply src/android/Print.java:728]: I was researching this and believe that it will be fine -- although there's similarity to the version name in this case, the version code of an Android app is independent from the user-displayed name, and must increase over time.

https://developer.android.com/studio/publish/versioning#versioningsettings
> The value is a positive integer so that other apps can programmatically evaluate it—to check an upgrade or downgrade relationship, for instance. You can set the value to any positive integer. However, make sure that each successive release of your app uses a greater value.

Perhaps PrintConnect has chosen to have their semver build portion match this version code, but even if they chose to switch over to a different naming scheme, the version code should be reliable for this purpose.

Comment (ssolbeck) [reply src/android/Print.java:728]: That said, a follow up PR to simplify the logic once more after the tablets are fully updated would be reasonable. That may involve some amount of coordination with GPM to ensure they are up-to-date.

**Comment by ssolbeck** [comment]
Have confirmed this build working locally with Android 12, PrintConnect 1.2.95, and an order app build using this PrintPlugin version.

**Pull request merged**
